hooks/go-imports/go-imports.go

```diff
@@ -16,62 +18,102 @@ import (
 	"github.com/inturn/pre-commit-gobuild/internal/helpers"
 )
 
+type lintError struct {
+	err  error
+	path string
+}
+
 func main() {
 	workDir, err := os.Getwd()
 	if err != nil {
 		log.Fatal(err)
 	}
 	dirs := helpers.DirsWith(workDir, "\\.go$")
 
+	errc := make(chan lintError, 10)
 	wg := &sync.WaitGroup{}
 
-	for _, dir := range dirs {
-		if !strings.Contains(dir, "/vendor/") {
-			files, err := ioutil.ReadDir(dir)
-			if err != nil {
-				log.Printf("error occured on read dir %s: %s", dir, err)
-			}
-			for _, f := range files {
-				if !strings.HasSuffix(f.Name(), ".go") {
-					continue
+	go func() {
+		for _, dir := range dirs {
+			if !strings.Contains(dir, "/vendor/") {
+				files, err := ioutil.ReadDir(dir)
+				if err != nil {
+					log.Printf("error occured on read dir %s: %s", dir, err)
+				}
+				for _, f := range files {
+					if !strings.HasSuffix(f.Name(), ".go") {
+						continue
+					}
+					wg.Add(1)
+					go func(d, name string) {
+						sortFileImports(filepath.Join(d, name), errc)
+						wg.Done()
+					}(dir, f.Name())
 				}
-				wg.Add(1)
-				go func(d, name string) {
-					sortFileImports(filepath.Join(d, name))
-					wg.Done()
-				}(dir, f.Name())
 			}
 		}
+		wg.Wait()
+		close(errc)
+	}()
+
+	var le []lintError
+	for lintErr := range errc {
+		log.Println(lintErr.path, lintErr.err)
+		le = append(le, lintErr)
+	}
+	if len(le) != 0 {
+		log.Println("files changed:", len(le))
+		os.Exit(1)
 	}
-	wg.Wait()
 }
 
-func sortFileImports(path string) {
+func sortFileImports(path string, errc chan<- lintError) {
 	fSet := token.NewFileSet()
 
 	f, err := parser.ParseFile(fSet, path, nil, parser.ParseComments)
 	if err != nil {
-		log.Println(err)
+		errc <- lintError{
+			err:  err,
+			path: path,
+		}
 		return
 	}
 
 	sortImports(f)
 
-	file, err := os.OpenFile(path, os.O_RDWR, 0666)
+	buf := &bytes.Buffer{}
+	if err := format.Node(buf, fSet, f); err != nil {
+		errc <- lintError{
+			err:  err,
+			path: path,
+		}
+		return
+	}
+
+	data, err := ioutil.ReadFile(path)
 	if err != nil {
-		log.Println(err)
+		errc <- lintError{
+			err:  err,
+			path: path,
+		}
 		return
 	}
 
-	if err := file.Truncate(0); err != nil {
-		log.Println(err)
```

Comment:
@yramanovich I think here we also need to have the same two modes - one for check and one for change. This way we will avoid some unexpected results because of the non-idempotent calls.

Reply:
@guntenbein added _--only-check_ flag